src/Compilers/CSharp/Portable/Parser/LanguageParser_Patterns.cs

```diff
@@ -131,7 +131,8 @@ private bool LooksLikeTypeOfPattern(SyntaxKind tk)
             return SyntaxFacts.IsPredefinedType(tk) ||
                 (tk == SyntaxKind.IdentifierToken && this.CurrentToken.ContextualKind != SyntaxKind.UnderscoreToken &&
                   (this.CurrentToken.ContextualKind != SyntaxKind.NameOfKeyword || this.PeekToken(1).Kind != SyntaxKind.OpenParenToken)) ||
-                LooksLikeTupleArrayType();
+                LooksLikeTupleArrayType() ||
+                IsFunctionPointerStart(); // We'll parse the function pointer, but issue an error in semantic analysis
```

Comment:
at thsi point, i would just break this huge expression into multiple statements.